### PR TITLE
edited check_item_status to make it more obj-oriented

### PR DESCRIPTION
**Pull request by bpalme**
This PR came out of the overdrive-integration.  Demian asked to separate this out.  The idea here is to make it easier to override the default functionality.   The base object is an ItemStatusHandler and defaults to the standard AJAX availability system.  A new handler can change the URL (for example) and add it to the handlers hash.  Here is some sample code:

```
//extend default handler and override any properties
var OdItemStatusHandler=Object.create(ItemStatusHandler);
  OdItemStatusHandler.url = '/Overdrive/getStatus';
  OdItemStatusHandler.itemStatusDelay = 200;
  OdItemStatusHandler.name = "odh";
  OdItemStatusHandler.itemStatusIds = [];
  OdItemStatusHandler.itemStatusEls = [];

//store the handlers in a "hash" obj
var handlers = {
    'ils':ItemStatusHandler,
    'overdrive':OdItemStatusHandler}
```
Then you can use the handler key in the template (like result-list.phtml):

```
<div class="media" data-handler-name="overdrive">
```

I haven't done a ton of testing, but I did put it back into a recent 5 install to make sure it still works with the current system.  I see Ere has a PR for formatting JS. This probably doesn't adhere to that, but I can try to work on that if you want.

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @bpalme! It appears for some reason that this PR currently has a conflict with master. Do you mind resolving that when time permits?

@crhallberg is currently on vacation, but since he's the lead developer on most of our Javascript at the moment, I'd like to hear his thoughts on this before reviewing it myself. I suspect his reaction will be positive, though!

I also agree that some style adjustments may be a good idea, especially in light of #1195... but that one is also waiting for @crhallberg to get back. Hopefully we can get the style adjustments merged to master shortly after Chris returns, then merge that into here, and get automated feedback on the required style fixes. Until then, I don't think you need to worry about it too much!

In other words, stay tuned for more feedback. :-)

File: themes/bootstrap3/js/check_item_statuses.js [hunk 1]
```diff
@@ -2,153 +2,176 @@
...
-    itemQueueAjax(id, $(ajaxItems[i]));
-  }
-  // Stop looking for a scroll loader
-  if (itemStatusObserver) {
-    itemStatusObserver.disconnect();
-  }
+    var ajaxItems = $(container).find('.ajaxItem');
+    for (var i = 0; i < ajaxItems.length; i++) {
+        var id = $(ajaxItems[i]).find('.hiddenId').val();
+        var handlerName = $(ajaxItems[i]).find('.handler').val();
```
Comment (crhallberg) [inline themes/bootstrap3/js/check_item_statuses.js]: Should we have `.handler` here be the same class as `.hiddenHandler` on new line 142 above?

File: themes/bootstrap3/js/check_item_statuses.js [hunk 2]
```diff
@@ -2,153 +2,176 @@
...
+            .fail( this.itemStatusFail)
+            .always(function(){
+                this.itemStatusRunning = false;
+            });
+    }//end runItemAjax
+};
+
+//store the handlers in a "hash" obj
+//add you own overridden handler here
+var handlers = {
```
Comment (crhallberg) [inline themes/bootstrap3/js/check_item_statuses.js]: Could we encapsulate this or give it a more specific name? I'd imagine `handlers` could be easily and accidentally overwritten by local custom JS.

**Comment by demiankatz** [comment]
@bpalme, @crhallberg: #1195 has been merged to master, so it should now be appropriate to fix the styles here. Can we do this automatically with eslint?

Also note that I resolved conflicts and merged master into this branch. I'd appreciate if you'd review my work and make sure I didn't break anything, though I tried to be careful! (note that I missed a lot of important details in 5c5c0c9 but hopefully fixed them in 16937f5).

**Comment by bpalme** [comment]
Thank you for doing the resolution.

File: themes/bootstrap3/js/check_item_statuses.js (hunk 1 above)
Comment (bpalme) [reply themes/bootstrap3/js/check_item_statuses.js]: I think it makes sense to make them the same.

File: themes/bootstrap3/js/check_item_statuses.js (hunk 2 above)
Comment (bpalme) [reply themes/bootstrap3/js/check_item_statuses.js]: It's a good idea.  I'll defer to you on which to do.

**Comment by demiankatz** [comment]
Thanks, both of you. I've just updated this again to merge in changes from #1204.

**Comment by bpalme** [comment]
I just realized that i missed some changes to this I think.  I need to add those changes and then double-check that it still works without the overdrive integration.

**Comment by demiankatz** [comment]
@bpalme, let me know if you need help! As noted above, I did try to bring this PR up to date with master after #1204. Was there something else that I missed?

**Comment by bpalme** [comment]
Sorry for the delay.  It turns out this is fine.  The problem was in my overdrive stuff.  So I think we are good to go on this.

**Comment by demiankatz** [comment]
Thanks for the confirmation! I'll take a closer look at this after the Summit next week.

**Comment by demiankatz** [comment]
@bpalme, this definitely looks to me like a step in the right general direction, and it's passing the automated test suite. Two questions, though:

1.) Why are there checks for `$item.find('.handler-name').val()` as well as `$item.data("handler-name")`? Is there a specific reason to support both of these use cases? If not, I would prefer to keep it simple and only support the data element case, which I think is more modern and straightforward. It looks like your Overdrive code is using data elements AND hidden fields with classes... not sure if that's necessary for some reason; just looking to simplify if we can!

2.) We're in the middle of a process to encapsulate all of our Javascript inside the top-level VuFind object for cleaner namespacing of code. For example, see cart.js. This basically amounts to wrapping most of the code in a function that returns an object exposing public methods, thus making everything else effectively private. Additionally, this object can expose an 'init' method that will be used for initialization (basically taking the place of the docReady event handler at the bottom of the file). I wonder if we should take care of encapsulating this file while we're refactoring it anyway, or whether that technique will interfere with any of the extensibility you have in mind. Any thoughts?

In any case, thanks for the work and for your patience (I know this is taking forever to review).

**Comment by bpalme** [comment]
That looks like an error to me.  I'll have to try it out and make sure it works with just one or the other.  

As for the encapsulation, I don't have any feelings about it as long as it works.

**Comment by demiankatz** [comment]
Thanks, @bpalme, I'll await your verdict on the hidden field/data attribute situation before doing further work.

Regarding encapsulation, I guess the biggest question is how much of a pain that will be for you to merge back into your Overdrive branch if I do it. If you'd like to see how it looks, I can take a few minutes and do the refactoring here... but if you think it will be a nuisance, we can leave this as-is for now and revisit that issue after all the Overdrive stuff is merged and complete.

**Comment by bpalme** [comment]
I would say go ahead and do the refactoring here and I'll make it work for Overdrive.  I don't think it will be that hard.  While I'm testing, I'll sort out that other issue as well.

**Comment by demiankatz** [comment]
Thanks, @bpalme, I've gone ahead and integrated everything into the VuFind namespace. See commit 9ccd346 for details (if you ignore whitespace diffs, it will make more sense -- most of the changes are just caused by indentation).

**Comment by bpalme** [comment]
The changes didn't affect the Overdrive integration. I just pasted in the new lines for Overdrive and it worked fine.  I'll commit that change shortly.  As for the other issue above I removed the second check and will just use the hidden field to designate the handler as we do the item ID.  So the template looks like this for a result item that gets its realtime availability from overdrive: 

```
<input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getOverdriveID())?>" class="hiddenId" />
<input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getSourceIdentifier())?>" class="hiddenSource" />
<input type="hidden" value="overdrive" class="handler-name" />
```

**Comment by demiankatz** [comment]
@bpalme, thanks for the update; sounds good. If you can commit the simplified data retrieval fix here as well, then I'll run the full test suite and make sure all the edge cases still work. If that passes, we can merge this first thing when 6.0 development begins.
